### PR TITLE
feat: UI fixes, L030 false-positive reduction, and noqa suppression

### PR DESCRIPTION
**Pull request by djdanielsson**
## Summary
- Fix multiple UI/UX issues: activity AI proposals, analytics descriptions, project violation counts
- Reduce L030 false positives by only flagging non-builtin modules that have a known builtin equivalent
- Add inline `# noqa: <rule_id>` suppression support for YAML content
- Fix nginx cache headers to prevent stale frontend bundles
- Fix session status bug (was using invalid enum value 4 instead of AWAITING_APPROVAL=1)

### UI & Gateway
- **Activity tab**: AI proposals now use `ExpandableSection` for consistent collapsible behavior
- **Analytics tab**: Added rule description column to AI Proposal Acceptance table
- **Projects page**: Fixed violations count mismatch (was showing paginated limit instead of true total) via new `project_violation_count` query
- **Project detail**: Increased violation fetch limit from 100 to 500

### L030 Rule
- Rule now checks whether the module's short name (e.g. `copy` from `community.general.copy`) exists in `ansible.builtin`
- Collection modules with no builtin counterpart (e.g. `community.general.timezone`, `amazon.aws.ec2_instance`) are no longer flagged
- Violation detail includes `builtin_alternative` field for actionable remediation guidance

### Inline noqa Suppression
- `parse_noqa()` extracts suppressed rule IDs from `# noqa: <rule_id>` comments in YAML
- Supports single (`# noqa: R108`) and multi-rule (`# noqa: R108, L030`) forms
- Integrated into `_evaluate_node()` in the graph scanner — suppressed rules are skipped during evaluation

### Nginx Cache Fix
- `index.html` served with `Cache-Control: no-cache, no-store, must-revalidate`
- `/assets/` served with immutable caching and `try_files $uri =404` (no SPA fallback for hashed assets)

### Bug Fixes
- Fixed `session.status = 4` (invalid) to `session.status = 1` (AWAITING_APPROVAL per proto enum) — proposals were silently dropped on session resume
- `ProposalsReady` now populates `tier` and `status` fields on initial emit (was only set on replay)
- CLI JSON mode no longer hangs waiting for interactive approval

## Test plan
- [x] `tox -e lint` passes
- [x] `tox -e unit` passes
- [x] L030 tests: 6 existing + 4 new tests pass (timezone, ec2_instance, no-builtin-equivalent cases)
- [x] Noqa tests: 6 new tests covering single/multi suppression, non-interference, parser

### CODE REVIEW

File: src/apme_engine/engine/graph_scanner.py [hunk 1]
```diff
@@ -163,6 +191,10 @@ def _evaluate_node(
 ) -> None:
     """Run all rules against a single node and append results to ``report``.
 
+    Rules suppressed via ``# noqa: <rule_id>`` in the node's YAML are
+    skipped.  Suppressed rules are still counted in diagnostics as
+    "suppressed" rather than "clean".
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/graph_scanner.py]: Good catch — the docstring claimed suppressed rules were counted in diagnostics but the implementation just skips them. Updated the docstring to accurately say they are skipped and not included in recorded results. Fixed in fc9f414.

File: src/apme_engine/daemon/primary_server.py [hunk 2]
```diff
@@ -1709,17 +1709,38 @@ async def _rescan_bridge(
...
-        if graph_report.ai_proposals:
-            proposals = self._build_graph_proposals(graph_report.ai_proposals)
-            for p in proposals:
+        # Yield AI proposals for human approval, or complete immediately.
+        # Build "declined" entries for AI-candidate violations the AI couldn't fix
+        # so the user sees them in the review panel.
+        proposed_proposals = self._build_graph_proposals(graph_report.ai_proposals) if graph_report.ai_proposals else []
+        proposed_rule_files: set[tuple[str, str]] = set()
+        for p in proposed_proposals:
+            proposed_rule_files.add((p.rule_id, p.file))
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/daemon/primary_server.py]: Valid — _build_graph_proposals() joins multiple rule_ids into a comma-separated string, so the set lookup wouldn't match individual IDs. Now splitting on comma and adding each individual rule_id to the set. Fixed in fc9f414.

File: src/apme_engine/daemon/primary_server.py [hunk 3]
```diff
@@ -1820,6 +1841,56 @@ def _build_graph_proposals(
...
+            rc_val = rc.value if hasattr(rc, "value") else str(rc) if rc else ""
+            if rc_val != RemediationClass.AI_CANDIDATE.value:
+                continue
+            rule_id = str(v.get("rule_id", ""))
+            file_path = str(v.get("file", ""))
+            if (rule_id, file_path) in proposed_rule_files:
+                continue
+            raw_line = v.get("line")
+            line_start = int(str(raw_line)) if raw_line is not None else 0
+            declined.append(
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/daemon/primary_server.py:1944]: Agreed — added defensive parsing that handles list/tuple (extracts first element), non-numeric strings, and any other unexpected types, with a blanket try/except falling back to 0. Fixed in fc9f414.

File: src/apme_engine/remediation/abbenay_provider.py [hunk 4]
```diff
@@ -102,6 +108,139 @@
...
+    {
+        "R108",  # Privilege escalation — often legitimate for service management
+        "R103",  # External command execution — may be necessary
+        "R104",  # Package install in task — may be intentional
+        "R101",  # Become user without become — could be inherited
+        "R105",  # Environment variable manipulation — may be required
+    }
+)
+
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/remediation/abbenay_provider.py]: Good catch on the duplication. Removed AI_REVIEWABLE_RULES from abbenay_provider.py — the canonical definition lives in partition.py. Fixed in fc9f414.

File: src/apme_engine/daemon/primary_server.py [hunk 5]
```diff
@@ -1768,17 +1768,41 @@ async def _rescan_bridge(
...
+
+        if all_proposals:
+            for p in proposed_proposals:
                 session.proposals[p.id] = p
-            session.ai_proposals = list(graph_report.ai_proposals)
-            session.status = 4  # AWAITING_APPROVAL
+            session.ai_proposals = list(graph_report.ai_proposals) if graph_report.ai_proposals else []
+
+            if proposed_proposals:
+                session.status = 4  # AWAITING_APPROVAL
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/daemon/primary_server.py]: Good catch — was using 4 which is not a valid SessionStatus enum value. The proto defines AWAITING_APPROVAL=1, PROCESSING=2, COMPLETE=3. Fixed to use 1. This also fixes session replay which checks status == 1. Fixed in c2723b9.

File: src/apme_engine/daemon/primary_server.py [hunk 6]
```diff
@@ -1768,17 +1768,41 @@ async def _rescan_bridge(
...
+            else:
+                session.status = 3  # COMPLETE
 
             yield SessionEvent(
-                proposals=ProposalsReady(proposals=list(proposals)),
+                proposals=ProposalsReady(proposals=all_proposals),
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/daemon/primary_server.py]: Valid — now populating tier=session.current_tier and status=session.status on the ProposalsReady event, matching the contract used by _session_replay_state and downstream consumers. Fixed in c2723b9.

File: src/apme_engine/engine/graph_scanner.py [hunk 7]
```diff
@@ -154,6 +158,30 @@ def load_graph_rules(
...
+    Returns:
+        Frozen set of suppressed rule IDs (empty if none found).
+    """
+    suppressed: set[str] = set()
+    for match in _NOQA_RE.finditer(yaml_lines):
+        for rule_id in match.group(1).split(","):
+            rid = rule_id.strip().upper()
+            if rid:
+                suppressed.add(rid)
+    return frozenset(suppressed)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/graph_scanner.py]: Good catch — now scanning line-by-line and stripping single/double-quoted strings before matching the noqa regex, so occurrences inside quoted YAML scalars (e.g. `shell: "echo # noqa: R108"`) are ignored. Fixed in 2c443f2.

File: src/apme_engine/daemon/primary_server.py [hunk 8]
```diff
@@ -1768,17 +1768,45 @@ async def _rescan_bridge(
...
+                proposals=ProposalsReady(
+                    proposals=all_proposals,
+                    tier=session.current_tier,
+                    status=session.status,
+                ),
             )
+
+            if session.status == 3:
+                async for event in self._session_build_result(session):
+                    yield event
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/daemon/primary_server.py]: Valid — restructured so ProposalsReady is only emitted when there are actual proposals to approve (proposed_proposals non-empty). Declined-only sets now skip the proposals event entirely and go straight to result, preventing double emission of _session_build_result and duplicate sink side-effects. Fixed in 2c443f2.

File: src/apme_engine/validators/native/rules/L030_non_builtin_use_graph.py [hunk 9]
```diff
@@ -28,51 +51,59 @@ class NonBuiltinUseGraphRule(GraphRule):
     """
 
     rule_id: str = "L030"
-    description: str = "Non-builtin module is used"
+    description: str = "Prefer ansible.builtin modules when available"
     enabled: bool = True
     name: str = "NonBuiltinUse"
-    version: str = "v0.0.2"
+    version: str = "v0.0.3"
     severity: Severity = Severity.LOW
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/validators/native/rules/L030_non_builtin_use_graph.py:58]: Good point — updated the description from "when available" to "when a builtin equivalent exists" to match the narrower trigger condition. Fixed in 2c443f2.

File: src/apme_engine/daemon/primary_server.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/daemon/primary_server.py]: Already fixed in c2723b9 — `session.status` is now set to `1` (AWAITING_APPROVAL) matching the proto enum. The code Copilot reviewed was from a prior commit.

File: src/apme_engine/daemon/primary_server.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/daemon/primary_server.py]: Already fixed in c2723b9 — `ProposalsReady` now populates `tier=session.current_tier` and `status=session.status`. The code Copilot reviewed was from a prior commit.

File: src/apme_engine/engine/graph_scanner.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/graph_scanner.py]: Acknowledged — technically valid but extremely low risk in practice. `parse_noqa()` operates on the raw YAML for a single node (not a full file), and `# noqa:` inside quoted strings in real Ansible content is an unlikely pattern. We accept this trade-off for simplicity. If this becomes a real issue we can switch to line-by-line parsing that skips quoted regions, but that adds complexity for a scenario we haven't seen in the wild.

File: src/apme_engine/daemon/primary_server.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/daemon/primary_server.py]: The current code (after c2723b9) handles this: when `all_proposals` contains only declined entries, `session.status` is set to `3` (COMPLETE) and `_session_build_result` is called immediately. The approval handler (`_session_apply_approved`) returns `0` early when `approved_ids` is empty and sets status to COMPLETE, making the path effectively idempotent. No double-emission occurs in practice.

File: src/apme_engine/validators/native/rules/L030_non_builtin_use_graph.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/validators/native/rules/L030_non_builtin_use_graph.py:58]: Good catch — updated the description from "Prefer ansible.builtin modules when available" to "Non-builtin module used when a builtin equivalent exists" across the rule, markdown docs, RULE_CATALOG, and generator script. Fixed in 3ac4cce.

File: src/apme_engine/engine/graph_scanner.py [hunk 10]
```diff
@@ -154,6 +158,36 @@ def load_graph_rules(
     }
 )
 
+_NOQA_RE = re.compile(r"#\s*noqa:\s*([A-Za-z0-9_,\t ]+)")
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/graph_scanner.py]: Good catch — anchored the regex with `(?:^|\s)` so `#` is only recognized at start-of-line or after whitespace. `echo#noqa: R108` no longer matches. Fixed in f37eb4a.

File: src/apme_gateway/db/queries.py [hunk 11]
```diff
@@ -312,6 +312,32 @@ async def project_violations(
...
+        select(Scan.scan_id).where(Scan.project_id == project_id).order_by(Scan.created_at.desc()).limit(1)
+    )
+    latest_result = await db.execute(latest_scan_stmt)
+    latest_scan_id = latest_result.scalar_one_or_none()
+    if latest_scan_id is None:
+        return 0
+
+    stmt = select(func.count()).select_from(Violation).where(Violation.scan_id == latest_scan_id)
+    result = await db.execute(stmt)
+    return int(result.scalar_one())
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/db/queries.py]: Valid — replaced the two-query approach with a single query on `Scan.total_violations`. One round-trip, no COUNT(*) on the violations table. Fixed in f37eb4a.

File: src/apme_gateway/api/router.py [hunk 12]
```diff
@@ -478,7 +478,7 @@ async def list_projects(
             trend_scans = await q.project_trend(db, proj.id, limit=5)
             last_scan_at = trend_scans[-1].created_at if trend_scans else None
             vt = _compute_violation_trend(trend_scans)
-            violations = await q.project_violations(db, proj.id)
+            violation_count = await q.project_violation_count(db, proj.id)
             items.append(
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py:482]: Added — `test_list_projects` now seeds with `scan_violations=5` and asserts `items[0]["total_violations"] == 5`. Fixed in f37eb4a.

File: src/apme_engine/engine/graph_scanner.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/graph_scanner.py]: Anchored the regex with `(?:^|\s)` so `#` must be at BOL or preceded by whitespace. This prevents false matches like `echo#noqa: R108` while still matching the standard `  # noqa: R108` form. Fixed in f37eb4a.

File: src/apme_gateway/db/queries.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/db/queries.py]: Replaced the two-query implementation with a single query on `Scan.total_violations`, avoiding the `COUNT(*)` on the violations table. Fixed in f37eb4a.

File: src/apme_gateway/api/router.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py:482]: Extended `test_list_projects` to seed violations and assert `total_violations` matches the seeded count. Fixed in f37eb4a.

File: src/apme_engine/engine/graph_scanner.py [hunk 13]
```diff
@@ -154,6 +158,36 @@ def load_graph_rules(
...
+    Args:
+        yaml_lines: Raw YAML text for a node.
+
+    Returns:
+        Frozen set of suppressed rule IDs (empty if none found).
+    """
+    suppressed: set[str] = set()
+    for line in yaml_lines.splitlines():
+        stripped = _QUOTED_RE.sub("", line)
+        for match in _NOQA_RE.finditer(stripped):
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/graph_scanner.py:186]: Acknowledged — relaxed the docstring to explicitly state that YAML `''` escape inside single-quoted scalars is not handled and is an accepted edge case. The regex approach covers the realistic patterns in Ansible content without the complexity of a full YAML-aware parser. Fixed in 37fc0e8.

File: tests/test_task_local_graph_rules_batch1.py [hunk 14]
```diff
@@ -699,3 +729,74 @@ def test_scan_avoid_implicit_violation(self) -> None:
...
+        assert node is not None
+        node.yaml_lines = "- name: Copy file\n  copy:  # noqa: L026\n    src: a\n    dest: /tmp/b\n"
+
+        rules: list[GraphRule] = [NonFQCNUseGraphRule()]
+        report = scan(g, rules)
+        for nr in report.node_results:
+            for rr in nr.rule_results:
+                meta = rr.rule
+                assert meta is None or meta.rule_id != "L026" or not rr.verdict
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_task_local_graph_rules_batch1.py:764]: Good catch — rewrote the test with a baseline scan (no `# noqa:`) that asserts L026 fires, then a suppressed scan asserting the violation is gone. The test now fails if suppression is broken or if the baseline doesn't produce a violation. Fixed in 37fc0e8.

File: tests/test_task_local_graph_rules_batch1.py [hunk 15]
```diff
@@ -699,3 +729,74 @@ def test_scan_avoid_implicit_violation(self) -> None:
...
+            "- name: Copy file  # noqa: L026, L030\n  community.general.copy:\n    src: a\n    dest: /tmp/b\n"
+        )
+
+        rules: list[GraphRule] = [NonFQCNUseGraphRule(), NonBuiltinUseGraphRule()]
+        report = scan(g, rules)
+        for nr in report.node_results:
+            for rr in nr.rule_results:
+                meta = rr.rule
+                assert meta is None or meta.rule_id not in ("L026", "L030") or not rr.verdict
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_task_local_graph_rules_batch1.py:812]: Same fix applied — added a baseline scan proving L026/L030 fire, then an explicit assertion that both are suppressed after adding `# noqa: L026, L030`. Fixed in 37fc0e8.

File: src/apme_engine/remediation/graph_engine.py [hunk 16]
```diff
@@ -378,13 +381,15 @@ async def remediate(
...
         remaining = graph.query_violations(status="open")
         fixed_violations = graph.query_violations(status="fixed")
+        ai_abstained = graph.query_violations(status="ai_abstained")
         step_diffs = graph.collect_step_diffs()
 
         return GraphFixReport(
             passes=passes,
             fixed=len(fixed_violations),
-            remaining_violations=remaining,
+            remaining_violations=remaining + ai_abstained,
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply src/apme_engine/remediation/graph_engine.py]: The ai_abstained code has been removed from this PR and moved to PR #274, where this concern is addressed — query_violations returns shallow copies and add_classification no longer touches resolution.

File: src/apme_engine/remediation/graph_engine.py [hunk 17]
```diff
@@ -589,13 +597,31 @@ async def _propose_one(
...
             )
 
+        # Mark violations on nodes where AI produced no fix at all.
+        from apme_engine.remediation.partition import normalize_rule_id  # noqa: PLC0415
+
+        for nid in node_keys:
+            if nid in proposed_node_ids:
+                continue
+            rule_ids = frozenset(normalize_rule_id(str(v.get("rule_id", ""))) for v in by_node[nid])
+            graph.abstain_violations(nid, rule_ids, pass_number=pass_num)
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply src/apme_engine/remediation/graph_engine.py]: The ai_abstained code has been removed from this PR and moved to PR #274, where this is addressed — failed_node_ids are tracked separately so AI exceptions are not misclassified as abstains.

File: src/apme_engine/engine/content_graph.py [hunk 18]
```diff
@@ -898,6 +901,43 @@ def decline_proposed(self, node_id: str) -> int:
...
+        for record in node.violation_ledger.values():
+            if record.status != "open":
+                continue
+            _, rule_id = record.key
+            if rule_id in rule_ids:
+                record.status = "ai_abstained"
+                record.fixed_in_pass = pass_number
+                record.violation["remediation_resolution"] = RemediationResolution.AI_ABSTAINED
+                count += 1
+        return count
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply src/apme_engine/engine/content_graph.py]: The ai_abstained code has been removed from this PR and moved to PR #274, where unit tests for the feature will be added.

File: src/apme_gateway/db/models.py [hunk 19]
```diff
@@ -172,7 +175,10 @@ class Violation(Base):
     remediation_class: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
     scope: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
     validator_source: Mapped[str] = mapped_column(Text, nullable=False, default="")
-    snippet: Mapped[str] = mapped_column(Text, nullable=False, default="")
+    original_yaml: Mapped[str] = mapped_column(Text, nullable=False, default="")
+    fixed_yaml: Mapped[str] = mapped_column(Text, nullable=False, default="")
+    co_fixes: Mapped[str] = mapped_column(Text, nullable=False, default="")
+    node_line_start: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply src/apme_gateway/db/models.py:181]: Added `_migrate_violations_table()` in `init_db` that inspects the violations table after `create_all` and runs `ALTER TABLE ADD COLUMN` for any of the four new columns (`original_yaml`, `fixed_yaml`, `co_fixes`, `node_line_start`) that are missing. Existing DBs are upgraded in-place without recreation. Fixed in 1b0f8c5.

File: frontend/src/theme.css [hunk 20]
```diff
@@ -80,18 +111,26 @@
...
-  font-size: 11px;
+  font-size: 10px;
   font-weight: 600;
+  white-space: nowrap;
+  line-height: 16px;
+}
+
+/* ── Severity badges ────────────────────────────────── */
+
+.apme-severity {
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply frontend/src/theme.css:131]: Restored the pill layout styles (`display: inline-flex`, padding, border-radius, font-size, etc.) on `.apme-severity` so callsites that don't include `apme-pill` still render correctly. Fixed in 1b0f8c5.

File: tests/test_remediation.py [hunk 21]
```diff
@@ -273,13 +266,16 @@ def test_remediation_resolution_is_str_enum(self) -> None:
         assert RemediationResolution.TRANSFORM_FAILED.value == "transform-failed"
         assert len(list(RemediationResolution)) == 10
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply tests/test_remediation.py:268]: The `AI_ABSTAINED` enum and proto changes have been moved to PR #274 (separate feature PR). The PR description for #266 has already been updated to remove references to `AI_ABSTAINED`. The test correctly asserts 10 members for this PR's scope.

File: frontend/src/components/ViolationOutputToolbar.tsx [hunk 22]
```diff
@@ -29,33 +36,47 @@ export function ViolationOutputToolbar({
...
+  const scopeRef = useRef<HTMLDivElement>(null);
+  const fixRef = useRef<HTMLDivElement>(null);
   const ruleRef = useRef<HTMLDivElement>(null);
 
   useEffect(() => {
     if (!filterMenuOpen) return;
     const handler = (e: MouseEvent) => {
-      const ref = filterMenuOpen === 'severity' ? sevRef : ruleRef;
-      if (ref.current && !ref.current.contains(e.target as Node)) {
+      const refs: Record<string, React.RefObject<HTMLDivElement | null>> = { severity: sevRef, scope: scopeRef, fix: fixRef, rule: ruleRef };
```
[the remark this answers is not in the document]
Comment (cidrblock) [reply frontend/src/components/ViolationOutputToolbar.tsx]: Imported `RefObject` directly from `react` and replaced the `React.RefObject` namespace reference. Fixed in 1b0f8c5.
